Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #30427 C++ tensor multi-dim indexing: more indexing tests
* **#30426 Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h**

This PR adds `assert_tensor_equal` and `assert_tensor_not_equal` to `test/cpp/api/support.h`, as better functions for testing whether two tensors are equal / not equal.

Differential Revision: [D18695900](https://our.internmc.facebook.com/intern/diff/D18695900)

Commit:
Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

kostmo commented:
## CircleCI build failures summary
As of commit 9dd14dc:
* **3/3** failures introduced in this PR



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=9dd14dca90d1c907ea721c87390b545c85a32314):

### 3 failures recognized by patterns:
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build
**Step:** Build
```
Nov 25 22:32:27 FAILED: test_api/CMakeFiles/test_api.dir/dataloader.cpp.o
```
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> binary_linux_manywheel_2_7mu_cpu_devtoolset7_build
**Step:** Build
```
Nov 25 22:32:52 /pytorch/test/cpp/api/support.h:94:38: error: no matching function for call to ‘at::Tensor::idx_put_(<brace-enclosed initializer list>, int)’
```
#### <img src='https://avatars0.githubusercontent.com/ml/7?s=12'/> binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_build
**Step:** Build
```
Nov 25 22:33:41 /pytorch/test/cpp/api/support.h:94:38: error: no matching function for call to ‘at::Tensor::idx_put_(<brace-enclosed initializer list>, int)’
```


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

yf225 commented on test/cpp/api/support.h:
```diff
@@ -49,6 +50,83 @@ inline bool pointer_equal(at::Tensor first, at::Tensor second) {
+        diff.idx_put_({nan_mask}, 0);
+        // inf check if allow_inf=true
+        if (allow_inf) {
+          auto inf_mask = torch::isinf(a);
```
Thanks for the catch! It should be added by https://github.com/pytorch/pytorch/pull/31099.

yf225 commented on test/cpp/api/support.h:
```diff
@@ -49,6 +50,83 @@ inline bool pointer_equal(at::Tensor first, at::Tensor second) {
+    }
+    b = b.to(a);
+
+    if ((a.dtype() == torch::kBool) != (b.dtype() == torch::kBool)) {
```
I feel that it might be a good idea to keep it in sync with the Python version to make maintenance easier https://github.com/pytorch/pytorch/blob/638e4ad8b95793876c0f5e82db58c18618fb6e61/test/common_utils.py#L804 If the Python version changes, we can change the C++ version correspondingly :D

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "[WIP] Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "[WIP] Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "[WIP] Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "[WIP] Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "[WIP] Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "[WIP] Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head

Commit:
Update on "Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h"

Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h

gh-metadata: pytorch pytorch 30426 gh/yf225/88/head
